README.md
## ConfigStorage
ConfigStorage is a library for the ESP32/ESP8266 Arduino platform to store configuration parameters in the file system using LittleFS or SPIFFS.
Using the ArduinoJson library the parameters are stored in memory as a JSON document and saved as a JSON file.

[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/Tost69/ConfigStorage/blob/master/LICENSE)

---
---

## Table of Contents

* [Changelog](changelog.md)
* [Prerequisites](#prerequisites)
* [Installation](#installation)
  * [Use Arduino Library Manager](#use-arduino-library-manager)
  * [Manual Install](#manual-install)
  * [VS Code & PlatformIO](#vs-code--platformio)
* [HOWTO Usage](#howto-usage)
* [Examples](#examples)
  * [ 1. ConfigStorage_ESP32](examples/ConfigStorage_ESP32)
  * [ 2. ConfigStorage_ESP8266](examples/ConfigStorage_ESP8266)
* [Debug](#debug)
* [Troubleshooting](#troubleshooting)
* [Issues](#issues)
* [TO DO](#to-do)
* [DONE](#done)
* [License](#license)
* [Copyright](#copyright)

---
---

## Prerequisites

1. [`Arduino IDE 1.8.19+` for Arduino](https://github.com/arduino/Arduino). [![GitHub release](https://img.shields.io/github/release/arduino/Arduino.svg)](https://github.com/arduino/Arduino/releases/latest)
2. [`ESP32 Core 2.0.2+`](https://github.com/espressif/arduino-esp32) for ESP32-based boards. [![Latest release](https://img.shields.io/github/release/espressif/arduino-esp32.svg)](https://github.com/espressif/arduino-esp32/releases/latest/)
3. [`ESP8266 Core 3.0.2+`](https://github.com/esp8266/Arduino) for ESP8266-based boards. [![Latest release](https://img.shields.io/github/release/esp8266/Arduino.svg)](https://github.com/esp8266/Arduino/releases/latest/). SPIFFS is deprecated from ESP8266 core 2.7.1+, to use LittleFS. 
4. [`LittleFS_esp32 v1.0.6+`](https://github.com/lorol/LITTLEFS) for ESP32-based boards using LittleFS with ESP32 core v1.0.5-. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/LittleFS_esp32.svg?)](https://www.ardu-badge.com/LittleFS_esp32). **Notice**: This [`LittleFS_esp32 library`](https://github.com/lorol/LITTLEFS) has been integrated to Arduino [ESP32 core v1.0.6+](https://github.com/espressif/arduino-esp32/tree/master/libraries/LITTLEFS) and you don't need to install it if using ESP32 core v1.0.6+
5. [`ArduinoJson v6.19.4+`](https://github.com/bblanchon/ArduinoJson) for embedded C++. To install, check [![arduino-library-badge](https://www.ardu-badge.com/badge/ArduinoJson.svg?)](https://www.ardu-badge.com/ArduinoJson).

---
---

## Installation

### Use Arduino Library Manager
The best and easiest way is to use `Arduino Library Manager`. Search for `ConfigStorage`, then select / install the latest version. You can also use this link [![arduino-library-badge](https://www.ardu-badge.com/badge/ConfigStorage.svg?)](https://www.ardu-badge.com/ConfigStorage) for more detailed instructions.

### Manual Install

1. Navigate to [ConfigStorage](https://github.com/Tost69/ConfigStorage) page.
2. Download the latest release `ConfigStorage-main.zip`.
3. Extract the zip file to `ConfigStorage-main` directory 
4. Copy the whole `ConfigStorage-main` folder to Arduino libraries' directory such as `~/Arduino/libraries/`.

### VS Code & PlatformIO:

1. Install [VS Code](https://code.visualstudio.com/)
2. Install [PlatformIO](https://platformio.org/platformio-ide)
3. Install [**ConfigStorage** library](https://registry.platformio.org/libraries/tost69/ConfigStorage) by using [Library Manager](https://registry.platformio.org/libraries/tost69/ConfigStorage/installation). Search for **ConfigStorage** in [Platform.io Author's Libraries](https://platformio.org/lib/search?query=author:%22Tost69%22)
4. Use included [platformio.ini](platformio/platformio.ini) file from examples to ensure that all dependent libraries will installed automatically. Please visit documentation for the other options and examples at [Project Configuration File](https://docs.platformio.org/page/projectconf.html)

---
---


changelog.md
## ConfigStorage

[![GitHub](https://img.shields.io/github/license/mashape/apistatus.svg)](https://github.com/Tost69/ConfigStorage/blob/master/LICENSE)

---
---

## Table of Contents

* [Changelog](#changelog)
  * [Releases v0.0.1](#releases-v001)
 
---
---

## Changelog

### Releases v0.0.1

1. Add LittleFS support for ESP8266
2. Add LittleFS support for ESP32


src/ESP_ConfigStore.h
/****************************************************************************************************************************
ESP_ConfigStore.h
For ESP32 and ESP8266 boards

Version Modified By   Date      Comments
------- -----------  ---------- -----------
0.0.1   Tom          20/04/2022 Initial coding
*****************************************************************************************************************************/

#pragma once

#ifndef ESP_ConfigStore_H
#define ESP_ConfigStore_H

#ifndef CONFIGSTORE_DEBUG
#define CONFIGSTORE_DEBUG       false
#endif // #ifndef CONFIGSTORE_DEBUG

#if defined(ARDUINO) && (ARDUINO >= 100)
#include <Arduino.h>
#else  // defined(ARDUINO) && (ARDUINO >= 100)
#include <WProgram.h>
#endif // #defined(ARDUINO) && (ARDUINO >= 100)

#ifndef ESP_CONFIG_STORE_VERSION
#define ESP_CONFIG_STORE_VERSION             "ESP_ConfigStore v0.0.1"
#define ESP_CONFIG_STORE_VERSION_MAJOR       0
#define ESP_CONFIG_STORE_VERSION_MINOR       0
#define ESP_CONFIG_STORE_VERSION_PATCH       1
#define ESP_CONFIG_STORE_VERSION_INT         0000001
#endif // #ifndef ESP_CONFIG_STORE_VERSION

#define ESP_CONFIGSTORE_VERSION         ESP_CONFIG_STORE_VERSION

//#define ESP_CS_USE_LITTLEFS    true
//#define ESP_CS_USE_SPIFFS      false

#ifdef ESP32
#if (!ESP_CS_USE_SPIFFS && !ESP_CS_USE_LITTLEFS)

#if (CONFIGSTORE_DEBUG)
#warning Neither LittleFS nor SPIFFS selected. Default to LittleFS
#endif // #if (CONFIGSTORE_DEBUG)

#ifdef ESP_CS_USE_LITTLEFS
#undef ESP_CS_USE_LITTLEFS
#define ESP_CS_USE_LITTLEFS      true
#endif // #ifdef ESP_CS_USE_LITTLEFS

#endif // #if (!ESP_CS_USE_SPIFFS && !ESP_CS_USE_LITTLEFS)
#endif // #ifdef ESP32

#ifdef ESP8266
#if (!ESP_CS_USE_SPIFFS && !ESP_CS_USE_LITTLEFS)

#if (CONFIGSTORE_DEBUG)
#warning Neither LittleFS nor SPIFFS selected. Default to LittleFS
#endif // #if (CONFIGSTORE_DEBUG)

#ifdef ESP_CS_USE_LITTLEFS
#undef ESP_CS_USE_LITTLEFS
#define ESP_CS_USE_LITTLEFS	true
#endif // #ifdef ESP_CS_USE_LITTLEFS

#endif // #if (!ESP_CS_USE_SPIFFS && !ESP_CS_USE_LITTLEFS)
#endif // #ifdef ESP8266

//default to use LITTLEFS (higher priority), then SPIFFS
#if ( ESP_CS_USE_LITTLEFS || ESP_CS_USE_SPIFFS )

#include <FS.h>

#ifdef ESP32

#if ESP_CS_USE_LITTLEFS

// Check cores/esp32/esp_arduino_version.h and cores/esp32/core_version.h
//#if ( ESP_ARDUINO_VERSION >= ESP_ARDUINO_VERSION_VAL(2, 0, 0) )  //(ESP_ARDUINO_VERSION_MAJOR >= 2)
#if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )

#if (CONFIGSTORE_DEBUG)
#warning Using ESP32 Core 1.0.6 or 2.0.0+
#endif // #if (CONFIGSTORE_DEBUG)

// The library has been merged into esp32 core from release 1.0.6
#include <LittleFS.h>

#define FileFS	LittleFS

#else // #if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )

#if (CONFIGSTORE_DEBUG)
#warning Using ESP32 Core 1.0.5-. You must install LITTLEFS library
#endif // #if (CONFIGSTORE_DEBUG)

// The library has been merged into esp32 core from release 1.0.6
#include <LITTLEFS.h>             // https://github.com/lorol/LITTLEFS

#define FileFS	LITTLEFS

#endif // #if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )

#else // #if ESP_CS_USE_LITTLEFS

#include "SPIFFS.h"
// ESP32 core 1.0.4 still uses SPIFFS
#define FileFS	SPIFFS

#endif // #if ESP_CS_USE_LITTLEFS

#else // #ifdef ESP32

// From ESP8266 core 2.7.1
#include <LittleFS.h>

#if ESP_CS_USE_LITTLEFS
#define FileFS    LittleFS
#else // #if ESP_CS_USE_LITTLEFS
#define FileFS   SPIFFS
#endif // #if ESP_CS_USE_LITTLEFS

#endif // #ifdef ESP32   

#endif // #if ( ESP_CS_USE_LITTLEFS || ESP_CS_USE_SPIFFS )   


//#include <Arduino.h>
//#include <ArduinoJson.h>
//#include "FS.h"
//#include <LITTLEFS.h>

//#define FileFS LITTLEFS
//#define FS_Name "LittleFS"

class ConfigStore
{
public:
  ConfigStore(const char *path)
  {
    if (!FileFS.begin())
    {
      Serial.println("LittleFS failed");
    }
    else
    {
      Serial.println("LittleFS started");

      fileName = path;
      readConfigFile();
    }
  };

  StaticJsonDocument<512> get()
  {
    return configDoc;
  };

  void set(DynamicJsonDocument Document)
  {
    configDoc = Document;
  };

  void initialize()
  {
    configDoc.clear();
  };

  void revert()
  {
    initialize();
    readConfigFile();
  };

  void save()
  {
    writeConfigFile();
  };

  void remove()
  {
    removeConfigFile();
  };

private:
  StaticJsonDocument<512> configDoc;
  const char *fileName;

  void readConfigFile()
  {
    if (FileFS.begin())
    {
      if (FileFS.exists(fileName))
      {
        Serial.println("Config file exists");

        File configFile = FileFS.open(fileName, "r");
        if (!configFile)
        {
          Serial.println("Reading config file failed");
        }
        else
        {
          Serial.println("Reading config file OK");

          deserializeJson(this->configDoc, configFile);
          configFile.close();
        }
      }
      else
      {
        Serial.println("Config file missing");
      }
    }
  };

  void writeConfigFile()
  {
    if (FileFS.begin())
    {
      File configFile = FileFS.open(fileName, "w");

      if (configFile)
      {
        serializeJson(configDoc, configFile);
        configFile.close();

        Serial.println("Saving config file OK");
      }
      else
      {
        Serial.println("Saving config file failed");
      }
    }
  };

  void removeConfigFile()
  {
    if (FileFS.begin())
    {
      if (FileFS.exists(fileName))
      {
        Serial.println("Config file exists");

        if (FileFS.remove(fileName))
        {
          Serial.println("Removing config file OK");
        }
        else
        {
          Serial.println("Removing config file failed");
        }
      }
      else
      {
        Serial.println("Config file missing");
      }
    }
  };
};


src/ConfigStorage.h
/****************************************************************************************************************************
ConfigStorage.h
for ESP32 and ESP8266 boards

Version Modified By   Date      Comments
------- -----------  ---------- -----------
0.0.1   Tom          20/04/2022 Initial coding
*****************************************************************************************************************************/

#pragma once

#ifndef ConfigStorage_H
#  define ConfigStorage_H

#  ifndef ConfigStorage_DEBUG
#    define ConfigStorage_DEBUG false
#  endif // #ifndef ConfigStorage_DEBUG

#  if defined(ARDUINO) && (ARDUINO >= 100)
#    include <Arduino.h>
#  else  // defined(ARDUINO) && (ARDUINO >= 100)
#    include <WProgram.h>
#  endif // #defined(ARDUINO) && (ARDUINO >= 100)

#  ifndef CONFIG_STORAGE_VERSION
#    define CONFIG_STORAGE_VERSION             "ConfigStorage v0.0.1"
#    define CONFIG_STORAGE_VERSION_MAJOR       0
#    define CONFIG_STORAGE_VERSION_MINOR       0
#    define CONFIG_STORAGE_VERSION_PATCH       1
#    define CONFIG_STORAGE_VERSION_INT         0000001
#  endif // #ifndef CONFIG_STORAGE_VERSION

#  define ConfigStorage_VERSION CONFIG_STORAGE_VERSION

//#define CS_USE_LITTLEFS    true
//#define CS_USE_SPIFFS      false

#  ifdef ESP32
#    include <ArduinoJson.h>
#    if (!CS_USE_SPIFFS && !CS_USE_LITTLEFS)

#      if (ConfigStorage_DEBUG)
#        warning Neither LittleFS nor SPIFFS selected. Default to LittleFS!
#      endif // #if (ConfigStorage_DEBUG)

#      ifdef CS_USE_LITTLEFS
#        undef CS_USE_LITTLEFS
#        define CS_USE_LITTLEFS true
#      endif // #ifdef CS_USE_LITTLEFS

#    endif // #if (!CS_USE_SPIFFS && !CS_USE_LITTLEFS)
#  endif // #ifdef ESP32

#  ifdef ESP8266
#    include <ArduinoJson.h>
#    if (!CS_USE_SPIFFS && !CS_USE_LITTLEFS)

#      if (ConfigStorage_DEBUG)
#        warning Neither LittleFS nor SPIFFS selected. Default to LittleFS!
#      endif // #if (ConfigStorage_DEBUG)

#      ifdef CS_USE_LITTLEFS
#        undef CS_USE_LITTLEFS
#        define CS_USE_LITTLEFS	true
#      endif // #ifdef CS_USE_LITTLEFS

#    endif // #if (!CS_USE_SPIFFS && !CS_USE_LITTLEFS)
#  endif // #ifdef ESP8266

// default to use LITTLEFS (higher priority), then SPIFFS
#  if ( CS_USE_LITTLEFS || CS_USE_SPIFFS )
#    include <FS.h>
#    ifdef ESP32
#      if CS_USE_LITTLEFS

//       check cores/esp32/esp_arduino_version.h and cores/esp32/core_version.h
#        if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )
#          if (ConfigStorage_DEBUG)
#            warning Using ESP32 Core 1.0.6 or 2.0.0+
#          endif // #if (ConfigStorage_DEBUG)

//         the library has been merged into esp32 core from release 1.0.6
#          include <LittleFS.h>
#          define FileFS	LittleFS

#        else // #if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )

#          if (ConfigStorage_DEBUG)
#            warning Using ESP32 Core 1.0.5-. You must install LITTLEFS library
#          endif // #if (ConfigStorage_DEBUG)

//         the library has been merged into esp32 core from release 1.0.6
#          include <LITTLEFS.h>             // https://github.com/lorol/LITTLEFS
#          define FileFS	LITTLEFS

#        endif // #if ( defined(ESP_ARDUINO_VERSION_MAJOR) && (ESP_ARDUINO_VERSION_MAJOR >= 2) )

#      else // #if CS_USE_LITTLEFS

//       ESP32 core 1.0.4 still uses SPIFFS
#        include "SPIFFS.h"
#        define FileFS	SPIFFS

#      endif // #if CS_USE_LITTLEFS

#    else // #ifdef ESP32

//     from ESP8266 core 2.7.1
#      include <LittleFS.h>

#      if CS_USE_LITTLEFS
#        define FileFS    LittleFS
#      else // #if CS_USE_LITTLEFS
#        define FileFS   SPIFFS
#      endif // #if CS_USE_LITTLEFS

#    endif // #ifdef ESP32   
#  endif // #if ( CS_USE_LITTLEFS || CS_USE_SPIFFS )   

//#include <Arduino.h>
//#include <ArduinoJson.h>
//#include "FS.h"
//#include <LITTLEFS.h>

//#define FileFS LITTLEFS
//#define FS_Name "LittleFS"

class ConfigStorage
{
public:
  ConfigStorage(const char *path)
  {
    if (!FileFS.begin())
    {
      Serial.println("LittleFS failed");
    }
    else
    {
      Serial.println("LittleFS started");

      fileName = path;
      readConfigFile();
    }
  };

  StaticJsonDocument<512> get()
  {
    return configDoc;
  };

  void set(DynamicJsonDocument Document)
  {
    configDoc = Document;
  };

  void initialize()
  {
    configDoc.clear();
  };

  void revert()
  {
    initialize();
    readConfigFile();
  };

  void save()
  {
    writeConfigFile();
  };

  void remove()
  {
    removeConfigFile();
  };

private:
  StaticJsonDocument<512> configDoc;
  const char *fileName;

  void readConfigFile()
  {
    if (FileFS.begin())
    {
      if (FileFS.exists(fileName))
      {
        Serial.println("Config file exists");

        File configFile = FileFS.open(fileName, "r");
        if (!configFile)
        {
          Serial.println("Reading config file failed");
        }
        else
        {
          Serial.println("Reading config file OK");

          deserializeJson(this->configDoc, configFile);
          configFile.close();
        }
      }
      else
      {
        Serial.println("Config file missing");
      }
    }
  };

  void writeConfigFile()
  {
    if (FileFS.begin())
    {
      File configFile = FileFS.open(fileName, "w");

      if (configFile)
      {
        serializeJson(configDoc, configFile);
        configFile.close();

        Serial.println("Saving config file OK");
      }
      else
      {
        Serial.println("Saving config file failed");
      }
    }
  };

  void removeConfigFile()
  {
    if (FileFS.begin())
    {
      if (FileFS.exists(fileName))
      {
        Serial.println("Config file exists");

        if (FileFS.remove(fileName))
        {
          Serial.println("Removing config file OK");
        }
        else
        {
          Serial.println("Removing config file failed");
        }
      }
      else
      {
        Serial.println("Config file missing");
      }
    }
  };
};

#endif // ConfigStorage_H